### PR TITLE
[SPARK-31234][SQL] ResetCommand should reset config to sc.conf only

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?
Currently, ResetCommand clear all configurations, including sql configs, static sql configs and spark context level configs.
for example:
```sql
spark-sql> set xyz=abc;
xyz abc
spark-sql> set;
spark.app.id local-1585055396930
spark.app.name SparkSQL::10.242.189.214
spark.driver.host 10.242.189.214
spark.driver.port 65094
spark.executor.id driver
spark.jars
spark.master local[*]
spark.sql.catalogImplementation hive
spark.sql.hive.version 1.2.1
spark.submit.deployMode client
xyz abc
spark-sql> reset;
spark-sql> set;
spark-sql> set spark.sql.hive.version;
spark.sql.hive.version 1.2.1
spark-sql> set spark.app.id;
spark.app.id <undefined>
```
In this PR, we restore spark confs to  RuntimeConfig after it is cleared

### Why are the changes needed?
reset command overkills configs which are static.
### Does this PR introduce any user-facing change?

yes, the ResetCommand do not change static configs now

### How was this patch tested?

add ut

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala [hunk 1]
```diff
@@ -170,7 +170,11 @@ object SetCommand {
 case object ResetCommand extends RunnableCommand with IgnoreCachedData {
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
-    sparkSession.sessionState.conf.clear()
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala]: Could you update [the comment](https://github.com/apache/spark/pull/28003/files) for the new behaivour?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala]: Also, I think we need to clearly define the behaivour of `RESET` in [the doc](https://github.com/apache/spark/blob/master/docs/sql-ref-syntax-aux-conf-mgmt-reset.md) like [hive](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Commands).

**Comment by yaooqinn** [comment]
also, cc @cloud-fan @gatorsmile thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -3258,7 +3258,9 @@ class SQLConf extends Serializable with Logging {
   }
 
   def clear(): Unit = {
```
Comment (yaooqinn) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: This API is exposed to users, and make the SQLConf hard to restore to its initial state after calling `spark.sessionState.conf.clear`

File: docs/sql-ref-syntax-aux-conf-mgmt-reset.md [hunk 3]
```diff
@@ -20,7 +20,7 @@ license: |
 ---
 
 ### Description
-Reset all the properties specific to the current session to their default values. After RESET command, executing SET command will output empty.
+Reset any properties specific to the current session which were set via the [SET](sql-ref-syntax-aux-conf-mgmt-set.html) command to their default values.
```
Comment (cloud-fan) [inline docs/sql-ref-syntax-aux-conf-mgmt-reset.md]: "any properties" -> "any runtime configs"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -3258,7 +3258,9 @@ class SQLConf extends Serializable with Logging {
...
+    val reserved = settings.asScala.filterKeys(staticConfKeys.contains).toMap
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: does `staticConfKeys` include the core configs?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala [hunk 5]
```diff
@@ -162,15 +162,20 @@ object SetCommand {
 }
 
 /**
- * This command is for resetting SQLConf to the default values. Command that runs
+ * This command is for resetting SQLConf to the default values. Any configurations that were set
+ * via [[SetCommand]] will get reset to default value Command that runs
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/SetCommand.scala]: a dot before `Command that runs`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: it doesn't. If a user calls it, it will make the SparkConf in SQLConf all gone

**Comment by yaooqinn** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `SQLConf` is internnal.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -3258,7 +3258,9 @@ class SQLConf extends Serializable with Logging {
...
     settings.clear()
+    settings.putAll(reserved.asJava)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: is this change necessary? static SQL configs can only be set with `SparkConf`. We can clear everything here, and the `RESET` command will add back the configs in `SparkConf`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]:
```scala
scala> spark.sessionState.conf.clear

scala> spark.sessionState.conf
res1: org.apache.spark.sql.internal.SQLConf = org.apache.spark.sql.internal.SQLConf@54e24511
```
show we make it private[sql] ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: yes. it is not necessary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: We don't. All classes under `org.apache.spark.sql.internal` are internal and users can still use. Spark marks things as internal by contract.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: OK.

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala [hunk 7]
```diff
@@ -45,7 +45,7 @@ class SQLConfSuite extends QueryTest with SharedSparkSession {
     // Clear the conf.
     spark.sessionState.conf.clear()
     // After clear, only overrideConfs used by unit test should be in the SQLConf.
-    assert(spark.conf.getAll === TestSQLContext.overrideConfs)
+    assert((spark.conf.getAll -- TestSQLContext.overrideConfs.keys).size < spark.conf.getAll.size)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: is this still needed?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 8]
```diff
@@ -1021,36 +1021,29 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
   }
 
   test("SET commands semantics using sql()") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: do we still need to change?

File: docs/sql-ref-syntax-aux-conf-mgmt-reset.md [hunk 9]
```diff
@@ -20,7 +20,7 @@ license: |
 ---
 
 ### Description
-Reset all the properties specific to the current session to their default values. After RESET command, executing SET command will output empty.
+Reset any runtime configurations specific to the current session which were set via the [SET](sql-ref-syntax-aux-conf-mgmt-set.html) command to their default values.
```
Comment (maropu) [inline docs/sql-ref-syntax-aux-conf-mgmt-reset.md]: We need to update the comment below, too?
```
-- Reset "all the properties" specific to the current session to their default values.
RESET;
```

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala [hunk 10]
```diff
@@ -115,6 +115,21 @@ class SQLConfSuite extends QueryTest with SharedSparkSession {
     }
   }
 
+  test("reset - static and spark conf") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: nit: the test name had better be self-descriptive.

File: docs/sql-ref-syntax-aux-conf-mgmt-reset.md (hunk 9 above)
Comment (yaooqinn) [reply docs/sql-ref-syntax-aux-conf-mgmt-reset.md]: good catch!

**Comment by yaooqinn** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala [hunk 11]
```diff
@@ -115,6 +115,21 @@ class SQLConfSuite extends QueryTest with SharedSparkSession {
...
 
+  test("reset will not change static sql configs and spark core configs") {
+    val conf = spark.sparkContext.getConf.getAll.toMap
+    val appName = conf.get("spark.app.name")
+    val driverHost = conf.get("spark.driver.host")
+    val master = conf.get("spark.master")
+    val warehouseDir = conf.get("spark.sql.warehouse.dir")
+    // ensure the conf here is not default value, and will not be reset to default value later
+    assert(warehouseDir.get.contains(this.getClass.getCanonicalName))
+    sql(s"reset")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: no need the `s`

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: and it's better to upper case the SQL keyword.

**Comment by maropu** [comment]
>> Does this PR introduce any user-facing change?
> yes, the ResetCommand do not change static configs now

btw, any chance to silently change behaviour of existing apps by this fix? We don't need to keep the current behaviour by a legacy config? cc: @gatorsmile @dongjoon-hyun @HyukjinKwon

**Comment by cloud-fan** [comment]
I think this is clearly a bug. If users set a static SQL config in the config file, this should not be affected by SET or  RESET.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by gatorsmile** [comment]
@yaooqinn Could you backport this to 2.4? and add it to the migration guide since it could impact the query results?

**Comment by yaooqinn** [comment]
> @yaooqinn Could you backport this to 2.4? and add it to the migration guide since it could impact the query results?

Hi @gatorsmile, why the migration guide should I put? In the backport or a followup to master?
